content/blog/firebase-cloud-functions.md

```diff
@@ -0,0 +1,361 @@
+      });
+
+      const preprocessedData = bills.map((item) => {
+        const newItem = { ...item };
+        if (newItem.time && newItem.time._seconds) {
+          // this function converts the timestamp object to time string, ignore if not needed
+          const dateDelist = new Date(newItem.time._seconds * 1000);
+          const dateInISTDelist = addMinutes(addHours(dateDelist, 5), 30);
+          newItem.time = format(dateInISTDelist, "dd-MM-yyyy hh:mm:ss aaa");
+        }
+
+        return newItem;
+      });
+
+      const csv = Papa.unparse(preprocessedData, {
+        quotes: false,
+        delimiter: ",",
+        header: true,
+        newline: "\r\n",
+        dynamicTyping: true,
+        skipEmptyLines: true,
+      });
+      await fs.outputFile(tempFilePath, csv);
+      await bucket.upload(tempFilePath, { destination: fileName });
+
+      const file = bucket.file(fileName);
+      await file.makePublic();
+
+      const publicUrl = `https://storage.googleapis.com/${bucket.name}/${fileName}`;
+
+      return await reportRef.update({
+        status: "complete",
+        downloadLink: publicUrl,
+        createdAt: admin.firestore.Timestamp.now(),
+      });
+    } catch (err) {
+      return console.log(err);
+    }
+  });
+```
+
+### Key Points
+
+<ul>
+  <li>
+ <strong>Trigger:</strong> The function is triggered by creating a new document in the reports collection.
+  </li>
+  <li>
+  <strong>Data Fetching:</strong> Retrieves all documents from the bills collection.
+  </li>
+  <li>
+  <strong>Date Manipulation:</strong> Converts Firestore timestamp fields to a more readable format using date-fns.
+  </li>
+  <li>
+  <strong>CSV Conversion:</strong> Uses Papaparse to convert the data to CSV format.
+  </li>
+  <li>
+<strong>File Handling:</strong> Creates a temporary file for the CSV, then uploads it to Firebase Storage.
+  </li>
+  <li>
+ <strong>Public Access:</strong> Makes the uploaded file publicly accessible.
+  </li>
+    <li>
+<strong>Update Firestore Document:</strong> Updates the report document with the CSV file’s download link and marks the report as complete.
+  </li>
+    <li>
+<strong>Error Handling:</strong> Includes try-catch for robust error management.
+  </li>
+</ul>
+
+### Conclusion
+
+In this tutorial, we’ve demonstrated how to automate the process of converting Firestore documents into a CSV format using Firebase Cloud Functions. This functionality is particularly useful for generating timely reports or exports, streamlining data management tasks without manual intervention.
+
+Below is a GIF that illustrates how the application behaves once integrated. This visual representation helps to understand the seamless flow from data trigger to CSV generation and retrieval.
+
+{{< figure src="/images/2024/firebase-cloud-function/firebase-cloud-function-2.gif" >}}
+
+For more detailed code, check out this GitHub repository: https://github.com/dixitt5/firestore-data-to-excel
```

Comment:
@dixitt5 let's add some more info to the readme of the repo

You can describe what's the purpose of the repo and how to use it and any other relevant info

Reply:
sure will do that.

Reply:
@2KAbhishek I have updated the readme here : https://github.com/dixitt5/firestore-data-to-excel
please check.